### PR TITLE
Feature/2020 05 30 terraform post

### PR DESCRIPTION
**Pull request by pietervincken**
### Thank you for your contribution! Please go through this small checklist to prepare your pull request to be reviewed as soon as possible :)
* [x] All authors were added to `_data/authors.yml` with a picture and an up-to-date bio
* [x] The blog post image is of ratio 500x293
  * Feel free to leave a comment if you don't know how to resize your image and someone will gladly help you out
* [x] The [**one sentence per line** method](https://raw.githubusercontent.com/brandon-rhodes/blog/master/texts/brandon/2012/one-sentence-per-line.rst) has been adopted
* [x] A spell check has been done
* [x] All links open in a new tab
  * Example: `[Amazing tech blog](https://ordina-jworks.github.io){:target="_blank" rel="noopener noreferrer"}`
* [x] The post's content properly resizes according to the window's size
  * Tip: Just shorten your window's horizontal length to easily test this
  * To have images properly resize use `class="image fit"`
  * To have YouTube videos properly resize, wrap the iframe into a `<div class="responsive-video">`
* [x] If a table of content is: check whether all links are working correctly

### Any specific remarks for the reviewers or anything you need help with?
\<enter text here>

### Checklist for the reviewers before merging
* [x] Publishing date of the post has been correctly set
  * This can be done by renaming the filename or by setting the date in the header section of the post

### CODE REVIEW

File: _posts/2020-05-30-terraform.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+1. Create reverse proxy configuration in reverse proxy repository
+1. Use a self-service portal to request access for Jenkins to update the configuration on the reverse proxy
+1. Roll-out the reverse proxy configuration using a Jenkins job
+1. Roll-out the environment configuration and application landscape to the new namespace through a Jenkins job that runs OC process and OC applies using the configuration.
+
+This sums up the best-case scenario.
+Since it requires multiple teams to perform disconnected tasks at different times, errors are frequent and slow to resolve.
+Next to that, there is a lot of manual copy and paste involved to sync configuration from different sources into different repositories and systems.
+Since any manual action, especially involving copying configuration between locations, is prone to errors, this introduces even more failure points.
+Combining the time it takes for tickets to be executed, copying the manual configuration to the correct locations and the debugging involved in getting the environment online results in lead times expressed in weeks.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:51]: Crazy, and not hard to believe with that many steps!

File: _posts/2020-05-30-terraform.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+An attempt was made at setting up the infrastructure with Terraform together with an ARM expert.
+Unfortunately, multiple walls were hit due to the nature of the corporate setup on Azure.
+One of the limitations that were hit, was the inability for ARM to make changes to resources that were provided by a central team.
+More specifically, using a VNET that wasn't managed by ARM to deploy an Azure Kubernetes Service proved to be difficult or even impossible.
+
+Using the CLI and scripts to create the setup was attempted as well, but this didn't fit the vision of a declarative setup for provisioning infrastructure.
+Re-applying the same scripts either broke the setup or created additional, unneeded resources.
+A clear no-go.
+
+As a third option, Terraform was investigated.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:74]: What is the difference here exactly compared to the first option? Seeing as Terraform was also investigated there.
I assume this option is about solely using Terraform?

File: _posts/2020-05-30-terraform.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+Looking online provided more questions than answers.
+A lot of documentation can be found about how to structure a specific module or how to create modules, but almost no resources discuss how to structure the modules into logical, reusable components.
+Until now!
+
+The setup was divided into two different categories: managed services wrappers and standard provisioning components.
+
+The first category is quite straight forward.
+For every managed service that is being used, a module is created.
+This module includes all the required Terraform resources for that service to operate.
+For example, the Key vault module contains the Azure Keyvault Terraform resource, but also the role assignments for the different Active Directory groups that require access to that Key vault.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:91]: I'd write it like this to distinguish between the generic term and the Azure keyed term.
```suggestion
For example, the key vault module contains the Azure Key Vault Terraform resource, but also the role assignments for the different Active Directory groups that require access to that key vault.
```

File: _posts/2020-05-30-terraform.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+It contains the Azure Container Registry (ACR) Terraform resource, some role assignments and the Azure Monitoring Diagnostics Settings for making sure the ACR logs are shipped into the correct logs analytics bucket.
+This abstraction allows for opinionated grouping of resources that are required for a managed service to operate.
+
+The second category is a grouping of modules of the first category.
+This set of modules are the ones that are actually provisioned during a deployment.
+Currently, only two of these modules exist a cluster module and a namespace module.
+
+The cluster module contains all modules required to set up the shared resources for all environments with a similar purpose.
+For example, all resources shared by all development environments.
+This module contains the Kubernetes cluster module, multiple resource group modules, a Keyvault module, some networking modules and some identity and role management modules.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:102]:
```suggestion
This module contains the Kubernetes cluster module, multiple resource group modules, a key vault module, some networking modules and some identity and role management modules.
```

File: _posts/2020-05-30-terraform.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+This module is used once per cluster, meaning once for development, acceptance and production respectively.
+
+The second module, namespace, is used to configure a namespace for a specific purpose: an environment for a development team, a specific testing environment or a rock-solid production environment.
+This module contains the DNS zone configuration, an API gateway module, another Keyvault module and a database module.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:106]:
```suggestion
This module contains the DNS zone configuration, an API gateway module, another key vault module and a database module.
```

File: _posts/2020-05-30-terraform.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+A lot of documentation can be found about how to structure a specific module or how to create modules, but almost no resources discuss how to structure the modules into logical, reusable components.
+Until now!
+
+The setup was divided into two different categories: managed services wrappers and standard provisioning components.
+
+The first category is quite straight forward.
+For every managed service that is being used, a module is created.
+This module includes all the required Terraform resources for that service to operate.
+For example, the Key vault module contains the Azure Keyvault Terraform resource, but also the role assignments for the different Active Directory groups that require access to that Key vault.
+Another example is the Container Registry module.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:92]: Same here as above.
```suggestion
Another example is the container registry module.
```

File: _posts/2020-05-30-terraform.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+It contains the Azure Container Registry (ACR) Terraform resource, some role assignments and the Azure Monitoring Diagnostics Settings for making sure the ACR logs are shipped into the correct logs analytics bucket.
+This abstraction allows for opinionated grouping of resources that are required for a managed service to operate.
+
+The second category is a grouping of modules of the first category.
+This set of modules are the ones that are actually provisioned during a deployment.
+Currently, only two of these modules exist a cluster module and a namespace module.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:98]: I'd add a colon here.
```suggestion
Currently, only two of these modules exist: a cluster module and a namespace module.
```

File: _posts/2020-05-30-terraform.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+The second module, namespace, is used to configure a namespace for a specific purpose: an environment for a development team, a specific testing environment or a rock-solid production environment.
+This module contains the DNS zone configuration, an API gateway module, another Keyvault module and a database module.
+
+These cluster and namespace modules are then used in a single Terraform module per cluster: so a single module for the development cluster together with all namespaces in that cluster.
+This makes making changes to the infrastructure as easy as running a single Jenkins job executing that module.
+Creating specific modules for development, acceptance and production also allows for the updates or upgrades of the configuration to be tested like application code before rolling it into production.
+
+## How can you beat us
+
+With this setup, lead-time for provisioning an environment went down from 6 weeks of ticketing magic to just a 90 minute Jenkins run.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:114]: Very cool!

File: _posts/2020-05-30-terraform.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+The decision to adopt Terraform still seems the correct one.
+Infrastructure upgrades have become more and more stable over the adoption period and the different modules have matured to the point where they barely change anymore.
+
+The work being done by the Azure Terraform provider community has helped tremendously.
+They release new versions of the provider every week.
+They aren't at feature parity with ARM, not by a long shot, but at the pace they are adding support for features, they will catch up very fast.
+The provider is also very stable and if an issue occurs (like [this](https://github.com/terraform-providers/terraform-provider-azurerm/issues/6525) AKS bug) it's fixed within the next release.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:121]: Open the link in a new tab to keep the reader with us. :)
```suggestion
The provider is also very stable and if an issue occurs (like [this](https://github.com/terraform-providers/terraform-provider-azurerm/issues/6525){:target="_blank" rel="noopener noreferrer"} AKS bug) it's fixed within the next release.
```

File: _posts/2020-05-30-terraform.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+## Finally, two golden tips when using Terraform
+
+The first is to start using remote state storage as soon as possible when using Terraform.
+It provides an easy way to get an accurate Terraform plan which in turn provides an accurate overview of the actions Terraform will execute during the roll-out.
+
+The second one is roll-out the modules often and validate their effects by running tests against them.
```
Comment (yannickdeturck) [inline _posts/2020-05-30-terraform.md:128]:
```suggestion
The second one is to roll out the modules often and validate their effects by running tests against them.
```

File: _posts/2020-05-30-terraform.md (hunk 2 above)
Comment (pietervincken) [reply _posts/2020-05-30-terraform.md:74]: Added

File: _posts/2020-06-02-terraform.md [hunk 11]
```diff
@@ -94,26 +99,47 @@ Another example is the container registry module.
...
+This seperation, in combination with the versioning has allowed us to test module updates, and upgrades of the configuration upfront.
+Similar to deploying application code to an acceptence or test environment before rolling it into production.
+
+## Where to draw a line, everything in terraform
+
+When we first started using terraform, we were tempted to configure the entire landscape using it.
+A good example was the API manager setup we were using.
+The Terraform setup included detailed configuration of the application it was hosting.
+Another example was a Kubernetes Terraform module we created, that beside infrastructure related setup for storage classes, was also creating service accounts for operators we were running.
+This leaking of concerns made it hard to maintain, but it also introduced duplication of configuration towards the application deployment.
```
Comment (yannickdeturck) [inline _posts/2020-06-02-terraform.md:126]: I think leakage sounds better?
```suggestion
This leakage of concerns made it hard to maintain, but it also introduced duplication of configuration towards the application deployment.
```

File: _posts/2020-06-02-terraform.md [hunk 12]
```diff
@@ -94,26 +99,47 @@ Another example is the container registry module.
...
+We decided to take a stricter approach, and consider two main responsibilities: orchestration and provisioning.
+Orchestration could be translated to an engineer that would take a screwdriver and setup all the required infrastructure.
```
Comment (yannickdeturck) [inline _posts/2020-06-02-terraform.md:128]:
```suggestion
Orchestration could be translated to an engineer that would take a screwdriver and set up all the required infrastructure.
```

File: _posts/2020-06-02-terraform.md [hunk 13]
```diff
@@ -94,26 +99,47 @@ Another example is the container registry module.
...
+The outcome of this work would be a the minimum needed setup to start using the services and configure them for actual use.
```
Comment (yannickdeturck) [inline _posts/2020-06-02-terraform.md:129]:
```suggestion
The outcome of this work would be the minimum needed setup to start using the services and configure them for actual use.
```

File: _posts/2020-06-02-terraform.md [hunk 14]
```diff
@@ -94,26 +99,47 @@ Another example is the container registry module.
...
+Provisioning is the work that comes after this, and doesn't require a screwdriver, or being near the box.
+We decided that the orchestration part, setting up the infrastructure, is where we use Terraform exclusively.
+The provisioning part, where we start to configure the infrastructure to work with our application, is where we use tools that are native to the application deployment.
+We didn't get there immediately though.
+In early iterations, we had split it into two Terraform runs, an orchestration run and a configuration run.
```
Comment (yannickdeturck) [inline _posts/2020-06-02-terraform.md:134]:
```suggestion
In early iterations, we had split it into two Terraform runs: an orchestration run and a configuration run.
```

File: _posts/2020-06-02-terraform.md [hunk 15]
```diff
@@ -94,26 +99,47 @@ Another example is the container registry module.
...
+In later iterations we were able to replace the configuration runs with components that were closer to the application landscape, which was a combination of privileged Kustomize deployment runs and building our own Kubernetes operators.
+A good rule of thumb is; if you can tie the component you are configuring to something that has meaning in the application domain, it should be managed in the application landscape.
+Another good indication that the application domain leaked into the infrastructure domain is when you have to re-orchestrate your infrastructure together with application changes.
 
 ## How can you beat us
 
 With this setup, lead-time for provisioning an environment went down from 6 weeks of ticketing magic to just a 90 minute Jenkins run.
-The decision to adopt Terraform still seems the correct one.
+The decision to adopt Terraform helped us to minimize human errors, and has enabled us to deploy new environments by a single push on a button.
```
Comment (yannickdeturck) [inline _posts/2020-06-02-terraform.md:142]: I think it's "of a button". :)
```suggestion
The decision to adopt Terraform helped us to minimize human errors, and has enabled us to deploy new environments by a single push of a button.
```

**Pull request merged**